### DISCUSSION TITLE
Belief Dependent Rewards

### DISCUSSION THREAD

**Discussion by josh0tt** [JuliaPOMDP/POMDPs.jl#387] [Q&A]
Are there any existing examples of a POMDPs.jl problem setup with belief-dependent rewards (rho-POMDP)? Currently, I have setup my problem as a belief MDP so that my reward function can make use of the belief.

**Comment by johannes-fischer** [comment]
I have worked with belief-dependent rewards in POMDPs.jl here https://github.com/johannes-fischer/icml2020_ipft, but it is not up to date with the current POMDPs.jl version. I was using a particle belief and a [belief MDP wrapper](https://github.com/johannes-fischer/icml2020_ipft/blob/master/IPFT/src/core/information_mdp.jl) to compute the belief-based reward from a particle belief.

Today, I would probably do it differently and introduce `abstract type ρPOMDP{S, A, O, B} end`. Downside is that then all methods in POMDPs.jl that are defined for `MDP`/`POMDP` are not available for the new type and would have to be reimplemented. I am not sure, if it is feasible or advisable to make `ρPOMDP` a subtype of `MDP{B,A}` or `POMDP{S,A,O}`, which would probably allow to reuse some methods without copy&paste-ing code for `ρPOMDP`. But declaring it as `MDP{B,A}` could be problematic since the transition should still be for states, not only for beliefs, and `POMDP{S,A,O}` might be undesired since a `POMDP` by definition has a state-based reward.

**Reply by zsunberg** [reply]
Yeah, this is a bit tricky.

Two general pieces of wisdom:
1. The more I have used Julia, the more I have moved away from abstract types.
2. Sometimes it's helpful to think about concrete tasks that you want to be able to do before jumping in to thinking about the code.

One side note: I think this is actually a design flaw of POMDPs.jl because we linked together the objective and the probabilistic models too tightly. Really, you should be able to take a transition model, observation model, and objective, and compose them arbitrarily. Someday, I think we should create DynamicDecisionNetworks.jl that allows one to do this.

In light of wisdom item (2), what is the task that you want to do? Some examples might be
- "take an existing POMDP add a belief-dependent reward function, and solve it using an MDP solver (e.g. MCTS)"
- "develop a rho-POMDP solver and make it easy for others to use"

once we have narrowed down the task, I might have more to say about how to design code for it.

**Reply by josh0tt** [reply]
Thank you both @johannes-fischer @zsunberg for the responses. I've been looking into @johannes-fischer implementation some more. I would say the task that I want to do is exactly the first bullet, "take an existing POMDP add a belief-dependent reward function, and solve it using an MDP solver (e.g. MCTS)"

My first pass was to just set up the MDP with the true state and the belief in the state struct. Then in the reward function, I can just take the expected reward over the belief (as well as including any additional belief entropy penalization terms). 

Is this what you would recommend? 

Right now I am keeping the true state around only to determine the true performance during the evaluation (printout the true reward for taking an action), but I guess this is unnecessary since during the tree search / rollouts only the belief is used.

**Reply by zsunberg** [reply]
Have you looked at [`GenerativeBeliefMDP`](https://juliapomdp.github.io/POMDPs.jl/stable/POMDPTools/model/#POMDPTools.ModelTools.GenerativeBeliefMDP)?

**Reply by zsunberg** [reply]
GenerativeBeliefMDP does not take the expectation of the reward over the belief - it just takes the reward over one sample, but you could write something like BeliefMDP to use the expectation of the state reward or another function of the belief (and contribute it back :)). Let me know if you want to do that and I can help.

**Reply by zsunberg** [reply]
> Right now I am keeping the true state around only to determine the true performance during the evaluation (printout the true reward for taking an action), but I guess this is unnecessary since during the tree search / rollouts only the belief is used.

I don't think you should do this. Instead, I'd recommend something like the following:
```julia
m = YourPOMDP()
up = DiscreteUpdater(m) # or particle filter or whatever you want
belief_reward(b, a) = sum(pdf(b, s)*reward(m, s, a) for s in support(b)) # this is the mean reward, but you could use any reward.
bm = BeliefMDP(m, up, belief_reward) # you would write this, similar to GenerativeBeliefMDP
planner = solve(MCTS.DPWSolver(), bm) # IMPORTANT: the planner only knows about bm, not m

# evaluate with this. IMPORTANT: we are simulating with the original POMDP, but the planner is using the belief MDP
mean(simulate(RolloutSImulator(), m, planner, up) for _ in 1:100)
```

**Reply by josh0tt** [reply]
This looks much more structured and helpful than the work arounds I'm using right now. The idea is that `BeliefMDP` would use `GenerativeBeliefMDP` but just define its own custom reward? I'm not sure I follow the last point though. Why simulate with the original POMDP instead of the belief MDP?

**Reply by zsunberg** [reply]
> The idea is that BeliefMDP would use GenerativeBeliefMDP but just define its own custom reward?

Not exactly. BeliefMDP would be *similar to* GenerativeBeliefMDP, but it would not *use* GenerativeBeliefMDP (except, perhaps a few methods could be defined for `Union{BeliefMDP,GenerativeBeliefMDP}`).

> I'm not sure I follow the last point though. Why simulate with the original POMDP instead of the belief MDP?

If you simulate with the belief MDP, the result will be the accumulated reward with the information gathering bonus; IIUC you want to evaluate how much reward in the *original* POMDP you get when using the belief-space planner.

**Reply by josh0tt** [reply]
to follow up on this, I created a `BeliefMDP` structure for a project I was working on [here](https://github.com/sisl/SBO_AIPPMS/blob/main/Rover/GP_BMDP_Rover/belief_mdp.jl) which is largely based off of `GenerativeBeliefMDP`. Right now, I've isolated it just for that project but it could be brought into POMDPs.jl later on.

**Reply by zsunberg** [reply]
Thanks @josh0tt , please do remember to circle back and see if you can contribute it if you find it useful!
